File: sections/12_stacks_queues_deques/problems/c_shortest_subarray_at_least/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int solve_case(const vector<long long> &a, long long target) {
    vector<long long> pref(a.size() + 1, 0);
    for (int i = 0; i < (int)a.size(); ++i) {
        pref[i + 1] = pref[i] + a[i];
    }

    int ans = (int)a.size() + 1;
    deque<int> dq;
    for (int i = 0; i < (int)pref.size(); ++i) {
        while (!dq.empty() && pref[i] - pref[dq.front()] >= target) {
            ans = min(ans, i - dq.front());
            dq.pop_front();
        }
        while (!dq.empty() && pref[dq.back()] >= pref[i]) {
            dq.pop_back();
        }
        dq.push_back(i);
    }
    return ans == (int)a.size() + 1 ? -1 : ans;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int T;
    cin >> T;
    while (T--) {
        int n;
        long long target;
        cin >> n >> target;
        vector<long long> a(n);
        for (long long &x : a) cin >> x;
        cout << solve_case(a, target) << '\n';
    }

    return 0;
}

File: sections/12_stacks_queues_deques/problems/c_shortest_subarray_at_least/solution.py
"""Reference solution for C. Shortest Subarray At Least K."""

from collections import deque
import sys


def solve_case(a: list[int], target: int) -> int:
    pref = [0]
    for x in a:
        pref.append(pref[-1] + x)

    ans = len(a) + 1
    dq: deque[int] = deque()
    for i, value in enumerate(pref):
        while dq and value - pref[dq[0]] >= target:
            ans = min(ans, i - dq.popleft())
        while dq and pref[dq[-1]] >= value:
            dq.pop()
        dq.append(i)
    return -1 if ans == len(a) + 1 else ans


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    t = data[0]
    idx = 1
    out: list[str] = []
    for _ in range(t):
        n = data[idx]
        target = data[idx + 1]
        idx += 2
        a = data[idx:idx + n]
        idx += n
        out.append(str(solve_case(a, target)))
    sys.stdout.write("\n".join(out))


if __name__ == "__main__":
    main()

